// @framer/google.android-home/code/AndroidSystem.tsx
import * as React from "react";
import { PropertyControls, ControlType, Override, Data } from "framer";
import styled from 'styled-components';

// Declare types so props are defined for later
type Theme = 'dark' | 'light'
type Context = 'lockscreen' | 'launcher' | 'app'

interface Props {
  theme: Theme
  context: Context
  version: string
  width: number
  height: number
  systemcolor: string
  navigationcolor: string
  tintcolor: string
  carrier: string
  locale: string
}

interface State {
  time: string
  date: string
}

export class AndroidSystem extends React.Component<Props, State> {

  // Set default properties
  static defaultProps = {
    theme: "dark",
    width: 360,
    height: 720,
    version: "9",
    systemcolor: 'rgba(0,0,0,0.1)',
    navigationcolor: 'rgba(0,0,0,1)',
    tintcolor: 'rgba(26,113,113,.5)',
    context: "lockscreen",
    time: "10:00",
    date: "Thursday September 13",
    carrier: "Framer",
    locale: "en-US"
  }

  // Items shown in property panel
  static propertyControls: PropertyControls<Props> = {
    version: {
      type: ControlType.SegmentedEnum,
      options: ["8", "9"],
      optionTitles: ["8", "9"],
      title: "Style",
    },
    context: {
      type: ControlType.SegmentedEnum,
      options: ["lockscreen", "launcher", "app"],
      optionTitles: ["Lock", "Home", "App"],
      title: "Context",
    },
    theme: {
      type: ControlType.SegmentedEnum,
      options: ["dark", "light"],
      optionTitles: ["Dark", "Light"],
      title: "Background",
    },
    systemcolor: { type: ControlType.Color, title: "Status Color"},
    navigationcolor: { type: ControlType.Color, title: "Nav Color"},
    tintcolor: { type: ControlType.Color, title: "BG Tint"},
    carrier: { type: ControlType.String, title: "Carrier"},
    locale: { type: ControlType.String, title: "Locale"}

  }

  timer: number

  state = {
    time: "9:00",
    date: "Thursday September 13"
  }

  componentDidMount() {
    this.timer = window.setInterval(() => {
      this.updateTime()
    }, 1000)
  }

  componentWillUnmount() {
      window.clearInterval(this.timer)
  }

  updateTime = () => {
    let date = new Date(),
        locale = this.props.locale,
        dateOptions = { weekday: 'long', month: 'long', day: 'numeric' },
        timeOptions = { hour: '2-digit', minute: '2-digit', hour12: false },
        time = date.toLocaleString(locale, timeOptions),
        formattedDate = date.toLocaleDateString(locale, dateOptions);
    this.setState({
      time: time,
      date: formattedDate
     });
  }

  render() {
    return (
      <SystemBars
        context={this.props.context}
        tintcolor={this.props.tintcolor}
        theme={this.props.theme}>
        <StatusBar
          context={this.props.context}
          version={this.props.version}
          systemcolor={this.props.systemcolor}>
          <Time>{ this.props.context === "lockscreen" ? this.props.carrier : this.state.time }</Time>
          <StatusBarIcons
            theme={this.props.theme}>
            {systemPlaceholders}
          </StatusBarIcons>
        </StatusBar>
        <Lockscreen>
          <LockscreenTime
            context={this.props.context}
            theme={this.props.theme}>
            {this.state.time}
          </LockscreenTime>
          <LockscreenDate
            context={this.props.context}
            theme={this.props.theme}>
            {this.state.date}
          </LockscreenDate>
        </Lockscreen>
        <NavBar
          context={this.props.context}
          theme={this.props.theme}
          navigationcolor={this.props.navigationcolor}>
          <NavButton>
            {this.props.version === "8" ? navBack : this.props.version === "9" && this.props.context === "app" && navBackP }
          </NavButton>
          <NavButton>
          {(this.props.version === "8") ? navHome : <NavHomeP theme={this.props.theme}/> }
          </NavButton>
          <NavButton>
            {(this.props.version === "8") ? navRecents : '' }
          </NavButton>
        </NavBar>
      </SystemBars>
    );
  }
}

// Reusable styles
const fastOutSlowIn = 'cubic-bezier(0.4, 0.0, 1, 1)',
      linearOutSlowIn = 'cubic-bezier(0.0, 0.0, 0.2, 1)',
      fastOutLinearIn = 'cubic-bezier(0.4, 0.0, 1, 1)';

// Styled components
const SystemBars = styled.div<Partial<Props>>`
  color: ${props => props.theme === 'dark' ? '#fff' : 'rgba(0,0,0,.87)' };
  background: ${props => props.context === 'lockscreen' ? props.tintcolor : 'transparent' };
  transition: .25s all ${fastOutSlowIn};
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`

const StatusBar = styled.div<Partial<Props>>`
  height: 24px;
  display: flex;
  justify-content: ${ props => props.version === "8" ? props.context === "lockscreen" ? "space-between" : "flex-start" : "space-between" };
  flex-direction: ${ props => props.version === "8" ? props.context === "lockscreen" ? "row" : "row-reverse" : "row" };
  align-items: center;
  padding: 0 16px;
  position: relative;
  overflow: hidden;
  &::before {
    content: '';
    position: absolute;
    width: 100%;
    height: 24px;
    z-index: 1;
    transition: .2s all ${linearOutSlowIn};
    background: ${ props => props.context === "app" ? props.systemcolor : "transparent" };
    top: ${ props => props.context === "app" ? '0' : '24px' };
  }
`

const Time = styled.div`
  font: 500 14px/1em 'Roboto';
  color: inherit;
  margin-right: 8px;
`

const StatusBarIcons = styled.div<Partial<Props>>`
  font: 16px/1em 'Material Icons';
  fill: ${props => props.theme === 'dark' ? '#fff' : '#000'};
  letter-spacing: -8px;
  display: flex;
  align-items: center;
  flex-grow: 0;
  flex-shrink: 0;
  margin-left: 8px;
`

const NavBar = styled.div<Partial<Props>>`
  display: flex;
  justify-content: ${ props => props.version === '8' ? 'center' : 'space-evenly' };
  stroke: ${props => props.theme === 'dark' ? '#fff' : '#5F6368'};
  align-items: center;
  height: 48px;
  width: 100%;
  position: relative;
  overflow: hidden;
  &::before {
    content: '';
    position: absolute;
    width: 100%;
    height: 48px;
    z-index: 1;
    transition: .2s all ${linearOutSlowIn};
    background: ${ props => props.context === "app" ? props.navigationcolor : "transparent" };
    transform: ${ props => props.context === "app" ? 'translateY(0)' : 'translateY(-48px)' };
  }
`
const NavButton = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 48px;
  width: 48px;
  position: relative;
  z-index: 2;
`

const NavHomeP = styled.div<Partial<Props>>`
  display: block;
  width: 28px;
  height: 10px;
  border-radius: 5px;
  background: ${ props => props.theme === 'light' ? 'rgba(0,0,0,0.7)' : '#fff' };
  box-shadow: ${ props => props.theme === 'light' ? '0' : '0 1px 1px rgba(0,0,0,.24)' };
`

const Lockscreen = styled.div`

`

const LockscreenTime = styled.div<Partial<Props>>`
  font: 400 56px Roboto, sans-serif;
  text-align: center;
  transition: .2s all ${linearOutSlowIn};
  color: ${props => props.theme === 'dark' ? '#fff' : 'rgba(0,0,0,.87)' };
  transform: ${ props => props.context === "lockscreen" ? 'translateY(0)' : 'translateY(-96px)' };
  opacity: ${ props => props.context === "lockscreen" ? '1' : '0' };
`

const LockscreenDate = styled.div<Partial<Props>>`
  font: 500 14px Roboto, sans-serif;
  text-align: center;
  transition: .2s all ${linearOutSlowIn};
  color: ${props => props.theme === 'dark' ? '#fff' : 'rgba(0,0,0,.87)' };
  transform: ${ props => props.context === "lockscreen" ? 'translateY(0)' : 'translateY(-48px)' };
  opacity: ${ props => props.context === "lockscreen" ? '1' : '0' };
`

// SVGs
const navBack = (
  <svg width="15px" height="17px" viewBox="0 0 15 17">
    <g transform="translate(-72.000000, -607.000000)"  fill="transparent" strokeWidth="2">
      <path d="M85.7724316,608.113105 L73.2232624,615.367718 C72.9340741,615.534896 72.934034,615.46698 73.2232624,615.634182 L85.772432,622.888794 C86.0605714,623.055366 86,623.09043 86,622.755562 L85.9999997,608.246337 C85.9999997,607.911516 86.0606134,607.946509 85.7724316,608.113105 Z" id="back"></path>
    </g>
  </svg>
);

const navHome = (
  <svg width="16px" height="16px" viewBox="0 0 16 16">
    <g transform="translate(-172.000000, -608.000000)"  fill="transparent" strokeWidth="2">
      <circle id="home" cx="180" cy="616" r="7"></circle>
    </g>
  </svg>
);

const navRecents = (
  <svg width="16px" height="16px" viewBox="0 0 16 16">
    <g transform="translate(-273.000000, -608.000000)" fill="transparent" strokeWidth="2">
      <rect id="recent" x="274" y="609" width="14" height="14" rx="2"></rect>
    </g>
  </svg>
);

const navBackP = (
  <svg width="7px" height="12px" viewBox="0 0 7 12" version="1.1">
    <polyline id="path-1" points="7 1.0575 5.95111583 0 0 6 5.95111583 12 7 10.9425 2.10520723 6"></polyline>
  </svg>
);

const systemPlaceholders = (
  <svg viewBox="0 0 62 24" width="62px" height="24px">
    <g id="Group-6" transform="translate(8, 7)" fill="inherit">
      <rect x="18" y="0" width="10" height="10" rx="5"></rect>
      <polygon  transform="translate(40.000000, 5.000000) scale(1, -1) translate(-40.000000, -5.000000) " points="40 0 46 10 34 10"></polygon>
      <rect x="0" y="0" width="10" height="10"></rect>
    </g>
  </svg>
);
